### PR TITLE
[3.4.5] sparkplug: rpc/attribute

### PR DESCRIPTION
**Pull request by nickAS21**
## Pull Request description
- shared attributes test
- rpc

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: application/src/test/java/org/thingsboard/server/transport/mqtt/sparkplug/connection/AbstractMqttV5ClientSparkplugConnectionTest.java [hunk 1]
```diff
@@ -22,31 +22,36 @@
...
 import java.util.HashSet;
+import java.util.List;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static org.awaitility.Awaitility.await;
 import static org.thingsboard.server.transport.mqtt.util.sparkplug.MetricDataType.Int32;
+import static org.thingsboard.server.transport.mqtt.util.sparkplug.SparkplugMessageTypeSate.OFFLINE;
```
Comment (ashvayka) [inline application/src/test/java/org/thingsboard/server/transport/mqtt/sparkplug/connection/AbstractMqttV5ClientSparkplugConnectionTest.java:43]: typo in the class name. Should be 'SparkplugMessageTypeState'. Th best name is SparkplugConnectionState

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java [hunk 2]
```diff
@@ -395,11 +402,11 @@ private void handleSparkplugPublishMsg(ChannelHandlerContext ctx, String topicNa
                         break;
                     case NBIRTH:
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:403]: I propose to simplify this in the following way

```java
                    case NBIRTH:
                        sparkplugSessionHandler.setNodeBirthMetrics(sparkplugBProtoNode.getMetricsList());
                    case NCMD:
                    case NDATA:
                        sparkplugSessionHandler.onAttributesTelemetryProto(msgId, sparkplugBProtoNode, deviceName, sparkplugTopic);
                        break;
```

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java [hunk 3]
```diff
@@ -416,7 +423,7 @@ private void handleSparkplugPublishMsg(ChannelHandlerContext ctx, String topicNa
                     case DBIRTH:
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:423]: Does DBIRTH have the metrics as well?

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java [hunk 4]
```diff
@@ -794,12 +801,29 @@ private void processRpcSubscribe(List<Integer> grantedQoSList, String topic, Mqt
...
         registerSubQoS(topic, grantedQoSList, reqQoS);
     }
 
+    public void processAttributesRpcSubscribeSparkplugNode(List<Integer> grantedQoSList, MqttQoS reqQoS) {
+        transportService.process(TransportProtos.TransportToDeviceActorMsg.newBuilder()
+                .setSessionInfo(deviceSessionCtx.getSessionInfo())
+                .setSubscribeToAttributes(SUBSCRIBE_TO_ATTRIBUTE_UPDATES_ASYNC_MSG)
+                .setSubscribeToRPC(SUBSCRIBE_TO_RPC_ASYNC_MSG)
+                .build(), null);
+//        transportService.process(deviceSessionCtx.getSessionInfo(), TransportProtos.SubscribeToAttributeUpdatesMsg.newBuilder().build(), null);
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:816]: remove commented code.

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java [hunk 5]
```diff
@@ -738,7 +738,7 @@ private void deregisterSession(String deviceName, MqttDeviceAwareSessionContext
         log.debug("[{}] Removed device [{}] from the gateway session", sessionId, deviceName);
     }
 
-    public void stateSparkplugtSendOnTelemetry (TransportProtos.SessionInfoProto sessionInfo, String deviceName, SparkplugMessageTypeSate typeSate, long ts) {
+    public void sendSparkplugStateOnTelemetry(TransportProtos.SessionInfoProto sessionInfo, String deviceName, SparkplugMessageTypeSate typeSate, long ts) {
         TransportProtos.KeyValueProto.Builder keyValueProtoBuilder = TransportProtos.KeyValueProto.newBuilder();
         keyValueProtoBuilder.setKey(STATE.name());
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java:743]: The key name should be "sparkplugConnectionState"

Comment (ashvayka) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java:743]: Do you send the same state update for all devices when the node disconnects?

File: application/src/test/java/org/thingsboard/server/transport/mqtt/sparkplug/connection/AbstractMqttV5ClientSparkplugConnectionTest.java (hunk 1 above)
Comment (nickAS21) [reply application/src/test/java/org/thingsboard/server/transport/mqtt/sparkplug/connection/AbstractMqttV5ClientSparkplugConnectionTest.java:43]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java (hunk 2 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:403]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java (hunk 3 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:423]: Yes

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java (hunk 4 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:816]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java (hunk 5 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java:743]: Yes, Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java (hunk 2 above)
Comment (ashvayka) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:403]: Please review my suggestion carefully. You have missed: ```sparkplugSessionHandler.setNodeBirthMetrics(sparkplugBProtoNode.getMetricsList());```

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMessageType.java [hunk 6]
```diff
@@ -87,6 +87,17 @@ public static SparkplugMessageType parseMessageType(String type) throws Thingsbo
 		}
 		throw new ThingsboardException("Invalid message type: " + type, ThingsboardErrorCode.INVALID_ARGUMENTS);
 	}
+	public static String messageName(SparkplugMessageType type) throws ThingsboardException {
+		if (STATE.equals(type)) {
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMessageType.java:91]: ```return STATE.equals(type) ? "sparkplugConnectionState" : type.name();```

**Pull request merged**